fix: handle forward-delete properly in the amount text input

akinwale opened the pull request:
### Details
Forward-delete in the text input means deleting text to the right hand side of the current cursor placement. This PR fixes the issue with the cursor jumping around when a forward-delete operation occurs.

Additional notes regarding the `TextInputWithCurrencySymbol` refactor will be posted below.

### Fixed Issues
$ https://github.com/Expensify/App/issues/25578
PROPOSAL: https://github.com/Expensify/App/issues/25578#issuecomment-1686288431

### Tests
1. Open the global create menu and then select "Request Money"
2. Input an amount (e.g., "12345678").
3. Position the cursor in between 2 or more digits, e.g. between 3 and 4 or between 4 and 5.
4. Press the keyboard's dedicated forward-delete button (if one is present).
5. Verify that a digit to the right hand side of the cursor is deleted, and the cursor does not jump around.
6. On Mac, press the keyboard shortcut: Fn+Delete.
7. Repeat step 5.
8. On Mac, press the keyboard shortcut: Control+D.
9. Repeat step 5.
10. Place the cursor at the end of the amount input.
11. Press the Backspace or Delete (on Mac) key.
12. Verify that the cursor does not jump a digit to the left.

- [x] Verify that no errors appear in the JS console

### Offline tests
N/A

### QA Steps
Same as tests.

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/4160319/a1b6b51c-3a27-4bbf-9b7d-7d697e9c7ab2

</details>

<details>
<summary>Mobile Web - Chrome</summary>

![25578-android-chrome](https://github.com/Expensify/App/assets/4160319/633c6d66-5b4d-4bd6-b611-ca3fb1313e8e)

</details>

<details>
<summary>Mobile Web - Safari</summary>

![25578-ios-safari](https://github.com/Expensify/App/assets/4160319/b0e11661-769c-4408-be32-532c4f0f8f3c)

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/4160319/3bc7c315-a585-4c65-a1a9-1736b5ed5fa6

</details>

<details>
<summary>iOS</summary>

![25578-ios-native](https://github.com/Expensify/App/assets/4160319/627e9070-7014-4155-a08f-4e7bca9c0a92)

</details>

<details>
<summary>Android</summary>

![25578-android-native](https://github.com/Expensify/App/assets/4160319/fcc3f1e0-2160-4d73-b725-671acf4900bd)

</details>

akinwale commented:
@narefyev91 

I had to refactor `TextInputWithCurrencySymbol` by splitting it into an Android-platform specific file (`index.android.js`) and the the `index.js` for the other platforms. This is due to a change that was introduced with this [PR](https://github.com/Expensify/App/pull/24177) which skips selection changes for certain updates to the text input.

This led to the issue outlined in the steps below.
1. Take the following value: `12345678` and place the cursor in between 2 and 3 like so: `12|345678`.
2. Forward-delete two digits to the right of the cursor so we have the following result: `12|5678`
3. Place the cursor at the end of the input: `125678|`
4. Press backspace to remove the 8 in front of the cursor. Since the last selection change was skipped, this will cause the cursor to jump to the front of 7, or in some cases 6, instead of being at the end of 7, so we have `1256|7` instead of `12567|`.

Since the [issue](https://github.com/Expensify/App/issues/22799) that the mentioned PR fixes only happens on Android, and the [related proposal](https://github.com/Expensify/App/issues/22799#issuecomment-1658728401) mentioned creating an Android-platform specific file, I have gone ahead to do this in this PR so that other platforms can work as expected.

Forward-delete is a no-op on Android, so this regression will never happen on Android even with a hardware keyboard connected (Delete key presses are not recognised).

akinwale committed:
handle forward-delete properly in the amount text input

narefyev91 commented:
@akinwale can you please sync your fork with latest main and also update branch and double check that code is still working

akinwale committed:
Merge remote-tracking branch 'upstream/main' into task-25578

akinwale committed:
handle Control-D keyboard shortcut on iOS

narefyev91 commented:
@akinwale can you please check ios-web

https://github.com/Expensify/App/assets/28352309/ecc24bd2-fc7e-47dc-8cf4-79f2c789775e

akinwale committed:
fix edge-case with iOS Safari's handling of the Control key for the keyboard shortcut

akinwale commented:
> @akinwale can you please check ios-web

@narefyev91 This seems to be a weird edge-case with the way iOS Safari handles the Control key for the keyboard shortcut. I have pushed a new commit with a specific fix for this. Everything tests OK on other platforms with this change.

narefyev91 commented:
## Reviewer Checklist
- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/28352309/647cdaac-e00e-4abb-b8d4-fdcdc62bd12a




</details>

<details>
<summary>Mobile Web - Chrome</summary>

![android-web](https://github.com/Expensify/App/assets/28352309/2b7c9077-54c0-415f-9b05-270574bae8ee)



</details>

<details>
<summary>Mobile Web - Safari</summary>


https://github.com/Expensify/App/assets/28352309/ac8c8696-4d34-4b7f-ba46-2c58f571203d




</details>

<details>
<summary>Desktop</summary>


https://github.com/Expensify/App/assets/28352309/d637640d-ae52-46bc-96f1-9bd402878460




</details>

<details>
<summary>iOS</summary>


https://github.com/Expensify/App/assets/28352309/a1affce7-8bc9-4ea7-9593-636e5472b644




</details>

<details>
<summary>Android</summary>

![android](https://github.com/Expensify/App/assets/28352309/29a3f563-cb64-4ffa-8df6-c0871bb524e3)



</details>

narefyev91 reviewed (approved):
LGTM! Nice and complex finding!
🎀 👀 🎀 C+ reviewed

marcaaron commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -171,6 +176,22 @@ function MoneyRequestAmountForm({amount, currency, isEditing, forwardedRef, onCu
+     */
+    const textInputKeyPress = ({nativeEvent}) => {
+        const key = nativeEvent.key.toLowerCase();
+        if (Browser.isMobileSafari() && key === 'control') {
```
`'control'` -> `CONST.PLATFORM_SPECIFIC_KEYS.CTRL.DEFAULT` ?

akinwale replied:
Make sense. I've seen these key constants in the past, but then I forgot all about them. It's fixed now.

akinwale committed:
use a constant to match for the 'control' key press

akinwale committed:
fix merge conflict

akinwale commented:
Retested on all platforms OK after fixing the merge conflict.

akinwale committed:
fix more merge conflicts

akinwale committed:
tidy up TextInputWithCurrencySymbol refactor

akinwale committed:
use anonymous function for onSelectionChange

akinwale commented:
@marcaaron Bump on review. I have fixed additional merge conflicts and cleaned up DRY issues with the code. I have also tested again on all platforms, after the modifications, to make sure that everything works as expected. Thanks.

akinwale committed:
fix merge conflict

akinwale commented:
Fixed another merge conflict and retested on all platforms. At least, we're all synced with the latest main now.

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6019259194) to staging by https://github.com/marcaaron in version: 1.3.59-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|failure ❌ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6031548810) to staging by https://github.com/marcaaron in version: 1.3.60-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6031548810) to staging by https://github.com/marcaaron in version: 1.3.60-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6031542575) to production by https://github.com/luacmartins in version: 1.3.59-5 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/6043148116) to production by https://github.com/luacmartins in version: 1.3.60-3 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅
